File: lab4_6_1/1_1.py
import heapq
import sys

def best_first_search(n, adj, start, target, h):
    # Min-priority queue: stores (heuristic_value, node_id)
    pq = [(h[start], start)]
    
    parent = [-1] * (n + 1)
    vis = [False] * (n + 1)
    vis[start] = True

    while pq:
        h_val, u = heapq.heappop(pq)

        if u == target:
            # Reconstruct path
            path = []
            curr = target
            while curr != -1:
                path.append(curr)
                curr = parent[curr]
            path.reverse()

            print("Path found: " + " -> ".join(map(str, path)))
            return

        if u in adj:
            for v in adj[u]:
                if not vis[v]:
                    vis[v] = True
                    parent[v] = u
                    heapq.heappush(pq, (h[v], v))
                    
    print("Target unreachable")

def main():
    try:
        input_data = sys.stdin.read().split()
        if not input_data:
            return
        it = iter(input_data)
        
        n = int(next(it))
        m = int(next(it))
        
        adj = {}
        for _ in range(m):
            u = int(next(it))
            v = int(next(it))
            if u not in adj: adj[u] = []
            if v not in adj: adj[v] = []
            adj[u].append(v)
            adj[v].append(u)

        start = int(next(it))
        target = int(next(it))

        h = []
        for _ in range(n + 1):
            h.append(int(next(it)))

        best_first_search(n, adj, start, target, h)
    except (StopIteration, ValueError, EOFError):
        pass

if __name__ == "__main__":
    main()

File: lab4_6_1/1_1.cpp
#include <bits/stdc++.h>
using namespace std;

void bestFirstSearch(int n, vector<vector<int>>& adj, int start, int target, vector<int>& h) {
    // Min-priority queue: stores {heuristic_value, node_id}
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
    
    vector<int> parent(n + 1, -1);
    vector<bool> vis(n + 1, false);

    pq.push({h[start], start});
    vis[start] = true;

    while (!pq.empty()) {
        int u = pq.top().second;
        pq.pop();

        if (u == target) {
            // Reconstruct path using a simple vector
            vector<int> path;
            for (int v = target; v != -1; v = parent[v]) path.push_back(v);
            reverse(path.begin(), path.end());

            cout << "Path found: ";
            for (int i = 0; i < path.size(); i++) 
                cout << path[i] << (i == path.size() - 1 ? "" : " -> ");
            cout << endl;
            return;
        }

        for (int v : adj[u]) {
            if (!vis[v]) {
                vis[v] = true;
                parent[v] = u;
                pq.push({h[v], v});
            }
        }
    }
    cout << "Target unreachable\n";
}

int main() {
    int n, m;
    cin >> n >> m;
    vector<vector<int>> adj(n + 1);
    for (int i = 0; i < m; i++) {
        int u, v; cin >> u >> v;
        adj[u].push_back(v); adj[v].push_back(u);
    }

    int start, target;
    cin >> start >> target;

    vector<int> h(n + 1);
    for (int i = 0; i <= n; i++) cin >> h[i];

    bestFirstSearch(n, adj, start, target, h);
    return 0;
}